File: Python/RR.py
def findWaitingTime(processes, bt, wt, quantum):
    n = len(processes)
    rem_bt = [0] * n
    for i in range(n):
        rem_bt[i] = bt[i]
    t = 0
    while True:
        done = True
        for i in range(n):
            if rem_bt[i] > 0:
                done = False
                if rem_bt[i] > quantum:
                    t += quantum
                    rem_bt[i] -= quantum
                else:
                    t += rem_bt[i]
                    wt[i] = t - bt[i]
                    rem_bt[i] = 0
        if done:
            break
    return t  # Total running time

def findTurnAroundTime(processes, bt, wt, tat):
    n = len(processes)
    for i in range(n):
        tat[i] = bt[i] + wt[i]

def findavgTime(processes, bt, quantum):
    n = len(processes)
    wt = [0] * n
    tat = [0] * n
    total_time = findWaitingTime(processes, bt, wt, quantum)
    findTurnAroundTime(processes, bt, wt, tat)

    print("Processes    Burst Time     Waiting Time    Turn-Around Time")
    total_wt = 0
    total_tat = 0
    for i in range(n):
        total_wt += wt[i]
        total_tat += tat[i]
        print(" ", processes[i], "\t\t", bt[i], "\t\t", wt[i], "\t\t", tat[i])

    print("\nAverage waiting time = %.5f " % (total_wt / n))
    print("Average turn around time = %.5f " % (total_tat / n))
    print("Total running time = %d" % total_time)
    print("Throughput = %.5f processes per time unit" % (n / total_time))

if __name__ == "__main__":
    proc = [1, 2, 3]
    burst_time = [10, 5, 8]
    quantum = 2
    findavgTime(proc, burst_time, quantum)

File: C#/RR.cs
using System;
using System.Diagnostics;

public class RR
{
	static void FindWaitingTime(int[] processes, int[] bt, int[] wt, int quantum)
	{
		int n = processes.Length;
		int[] rem_bt = new int[n];
		Array.Copy(bt, rem_bt, n);
		int t = 0;
		while (true)
		{
			bool done = true;
			for (int i = 0; i < n; i++)
			{
				if (rem_bt[i] > 0)
				{
					done = false;
					if (rem_bt[i] > quantum)
					{
						t += quantum;
						rem_bt[i] -= quantum;
					}
					else
					{
						t += rem_bt[i];
						wt[i] = t - bt[i];
						rem_bt[i] = 0;
					}
				}
			}
			if (done)
				break;
		}
	}

	static void FindTurnAroundTime(int[] processes, int[] bt, int[] wt, int[] tat)
	{
		int n = processes.Length;
		for (int i = 0; i < n; i++)
			tat[i] = bt[i] + wt[i];
	}

	static void _RR(int[] processes, int[] bt, int quantum)
	{
		int n = processes.Length;
		int[] wt = new int[n];
		int[] tat = new int[n];
		var stopwatch = Stopwatch.StartNew();
		FindWaitingTime(processes, bt, wt, quantum);
		FindTurnAroundTime(processes, bt, wt, tat);
		stopwatch.Stop();
		double total_wt = 0, total_tat = 0;
		for (int i = 0; i < n; i++)
		{
			total_wt += wt[i];
			total_tat += tat[i];
		}
		double avg_wt = total_wt / n;
		double avg_tat = total_tat / n;
		double throughput = n / stopwatch.Elapsed.TotalSeconds;
		Console.WriteLine("Processes Burst Time Waiting Time Turn-Around Time");
		for (int i = 0; i < n; i++)
			Console.WriteLine($" {processes[i]}        {bt[i]}        {wt[i]}          {tat[i]}");
		Console.WriteLine($"\nAverage waiting time = {avg_wt:F5}");
		Console.WriteLine($"Average turn around time = {avg_tat:F5}");
		Console.WriteLine($"Throughput = {throughput}");
		Console.WriteLine($"Running time = {stopwatch.Elapsed.TotalSeconds:F5} seconds");
	}

	public static void Main()
	{
		int[] processes = { 1, 2, 3 };
		int[] burst_time = { 10, 5, 8 };
		int quantum = 2;
		_RR(processes, burst_time, quantum);
	}
}
